In [1]:
import pandas as pd

df = pd.read_excel("../data/raw/ApexPlanet_DataAnalytics_Dataset.xlsx")
df.head()

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
0,ORD100002,2025-02-25,CUST5529,Customer_227,30.0,Female,Bengaluru,Rice,Grocery,7,2829.77,19808.39
1,ORD100003,2025-10-14,CUST3127,Customer_182,63.0,Male,Bengaluru,Book,Education,5,27906.16,139530.80
2,ORD100004,2025-05-13,CUST8887,Customer_487,62.0,Female,Bengaluru,Book,Education,8,37491.06,299928.48
3,ORD100005,2025-12-02,CUST2515,Customer_470,65.0,Female,Kolkata,Mobile,Electronics,9,28541.36,256872.24
4,ORD100006,2025-11-20,CUST4796,Customer_380,44.0,Male,Bengaluru,Rice,Grocery,10,14036.59,140365.90


In [2]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 1000
Columns: 12


In [3]:
df.columns

Index(['Order_ID', 'Order_Date', 'Customer_ID', 'Customer_Name', 'Age',
       'Gender', 'City', 'Product', 'Category', 'Quantity', 'Unit_Price',
       'Total_Sales'],
      dtype='object')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order_ID       1000 non-null   object 
 1   Order_Date     1000 non-null   object 
 2   Customer_ID    1000 non-null   object 
 3   Customer_Name  1000 non-null   object 
 4   Age            980 non-null    float64
 5   Gender         1000 non-null   object 
 6   City           987 non-null    object 
 7   Product        1000 non-null   object 
 8   Category       1000 non-null   object 
 9   Quantity       1000 non-null   int64  
 10  Unit_Price     1000 non-null   float64
 11  Total_Sales    1000 non-null   float64
dtypes: float64(3), int64(1), object(8)
memory usage: 93.9+ KB


In [5]:
df.isnull().sum()

Order_ID          0
Order_Date        0
Customer_ID       0
Customer_Name     0
Age              20
Gender            0
City             13
Product           0
Category          0
Quantity          0
Unit_Price        0
Total_Sales       0
dtype: int64

In [6]:
df.duplicated().sum()

np.int64(0)

In [7]:
df.describe()

,Age,Quantity,Unit_Price,Total_Sales
count,980.000000,1000.000000,1000.000000,1000.000000
mean,41.360204,5.435000,25486.783410,139399.439650
std,13.822597,2.838632,14179.402361,114100.051546
min,18.000000,1.000000,145.780000,437.340000
25%,30.000000,3.000000,13895.722500,47066.632500
50%,41.000000,5.000000,25398.740000,108594.025000
75%,54.000000,8.000000,37512.382500,203722.882500
max,65.000000,10.000000,49997.530000,493677.500000


In [8]:
df.describe(include="object")

,Order_ID,Order_Date,Customer_ID,Customer_Name,Gender,City,Product,Category
count,1000,1000,1000,1000,1000,987,1000,1000
unique,992,342,947,425,2,8,6,5
top,ORD100050,2025-04-16,CUST2515,Customer_359,Male,Patna,Mobile,Electronics
freq,9,10,3,8,511,135,184,354


In [9]:
df["City"].unique()

array(['Bengaluru', 'Kolkata', 'Hyderabad', 'Patna', 'Mumbai', 'Delhi',
       'Gaya', 'Pune', nan], dtype=object)

In [10]:
print(f"Start date: {df["Order_Date"].min()}")
print(f"Start date: {df["Order_Date"].max()}")

Start date: 2025-01-01
Start date: 2026-01-01


In [11]:
df['Category'].unique()

array(['Grocery', 'Education', 'Electronics', 'Fashion', 'Furniture'],
      dtype=object)

In [12]:
df["Product"].unique()

array(['Rice', 'Book', 'Mobile', 'Laptop', 'Shoes', 'Chair'], dtype=object)

In [13]:
for col in df.columns:
    print(col, df[col].nunique())

Order_ID 992
Order_Date 342
Customer_ID 947
Customer_Name 425
Age 48
Gender 2
City 8
Product 6
Category 5
Quantity 10
Unit_Price 1000
Total_Sales 1000


In [14]:
# Check Order_ID duplicates
df["Order_ID"].duplicated().sum()

np.int64(8)

In [15]:
duplicate_orders = df[df["Order_ID"].duplicated(keep=False)]
duplicate_orders.sort_values("Order_ID")

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
48,ORD100050,2025-10-03,CUST5523,Customer_432,NaN,Male,Pune,Chair,Furniture,9,30834.87,277513.83
118,ORD100050,2025-01-22,CUST3194,Customer_114,34.0,Male,Kolkata,Book,Education,2,25795.55,51591.10
238,ORD100050,2025-03-01,CUST8467,Customer_104,23.0,Male,Bengaluru,Book,Education,3,2860.83,8582.49
358,ORD100050,2025-02-11,CUST6467,Customer_181,24.0,Female,Pune,Rice,Grocery,7,45480.39,318362.73
478,ORD100050,2025-11-22,CUST8225,Customer_468,64.0,Male,Bengaluru,Rice,Grocery,8,38957.59,311660.72
598,ORD100050,2025-02-21,CUST8859,Customer_53,NaN,Male,NaN,Laptop,Electronics,7,20144.84,141013.88
718,ORD100050,2025-11-06,CUST7824,Customer_190,31.0,Female,Mumbai,Rice,Grocery,4,42642.59,170570.36
838,ORD100050,2025-08-03,CUST5585,Customer_289,35.0,Male,Kolkata,Chair,Furniture,9,28747.47,258727.23
958,ORD100050,2025-02-09,CUST7188,Customer_89,62.0,Male,Bengaluru,Mobile,Electronics,1,34619.77,34619.77


# Key Findings

## Dataset Overview
- Dataset contains **1000 records** and **12 columns**.
- Data includes customer demographics, product information, sales details, and transaction dates.
- The dataset covers transactions from **01-Jan-2025 to 01-Jan-2026**.

---

## Data Quality Assessment

### Missing Values

| Column | Missing Values |
|----------|----------|
| Age | 20 |
| City | 13 |

#### Observations
- The **Age** column contains 20 missing values.
- The **City** column contains 13 missing values.
- All other columns have complete records.

---

### Duplicate Analysis

#### Full Row Duplicates
- No completely duplicated rows were found in the dataset.

#### Order_ID Validation
A detailed investigation was performed on duplicate Order_ID values.

##### Findings
- The dataset contains **992 unique Order_IDs** out of **1000 records**.
- A total of **8 duplicate Order_ID occurrences** were identified.
- All duplicate records share the Order_ID **ORD100050**.
- These records belong to different customers, order dates, products, quantities, and sales amounts.

##### Business Concern
Normally, a single Order_ID should uniquely identify a transaction. The repeated use of the same Order_ID across multiple customers indicates a possible data-entry or system-generated inconsistency.

---

## Data Types

### Numerical Columns
- Age
- Quantity
- Unit_Price
- Total_Sales

### Categorical Columns
- Order_ID
- Customer_ID
- Customer_Name
- Gender
- City
- Product
- Category

### Date Column
- Order_Date is currently stored as an **object** datatype and requires verification before analysis.

---

## Numerical Insights

### Age
| Metric | Value |
|----------|----------|
| Minimum | 18 |
| Maximum | 65 |
| Mean | 41.36 |
| Median | 41 |

### Quantity
| Metric | Value |
|----------|----------|
| Minimum | 1 |
| Maximum | 10 |
| Mean | 5.44 |

### Unit Price
| Metric | Value |
|----------|----------|
| Minimum | 145.78 |
| Maximum | 49,997.53 |
| Mean | 25,486.78 |

### Total Sales
| Metric | Value |
|----------|----------|
| Minimum | 437.34 |
| Maximum | 493,677.50 |
| Mean | 139,399.44 |

---

## Categorical Insights

| Feature | Unique Values |
|----------|----------|
| Gender | 2 |
| City | 8 |
| Product | 6 |
| Category | 5 |

### Most Frequent Values
- Most common city: **Patna**
- Most common product: **Mobile**
- Most common category: **Electronics**
- Most common gender: **Male**

---

## Initial Observations

### Positive Findings
- Dataset contains complete sales-related information.
- No fully duplicated records were detected.
- Customer, product, and transaction information are well represented.
- Numerical and categorical features are clearly distinguishable.

### Data Quality Concerns
1. Missing values exist in **Age** and **City** columns.
2. Duplicate usage of **Order_ID (ORD100050)** was identified.
3. Order_Date is stored as an object datatype.
4. Sales calculations have not yet been validated.
5. Potential outliers may exist in sales-related variables and require further investigation.

---

## Conclusion

The dataset provides a comprehensive view of customer purchases across multiple cities and product categories. Initial profiling indicates that the dataset is largely complete, with only a small number of missing values and no fully duplicated records. However, duplicate Order_ID entries, missing demographic information, datatype inconsistencies, and potential outliers should be reviewed before proceeding with advanced analysis and visualization.
